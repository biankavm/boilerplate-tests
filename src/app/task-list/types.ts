import { Task } from '@/types/graphql'

export type TaskDTO = Omit<Task, 'id'>
