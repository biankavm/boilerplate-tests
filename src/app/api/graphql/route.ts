import { NextRequest, NextResponse } from 'next/server'

let tasks = [
  {
    id: '1',
    title: 'Tarefa exemplo 1',
    completed: false,
    createdAt: new Date().toISOString()
  },
  {
    id: '2',
    title: 'Tarefa exemplo 2',
    completed: true,
    createdAt: new Date().toISOString()
  }
]

let users = [
  {
    id: '1',
    name: 'Jo√£o Silva',
    email: 'joao@example.com',
    createdAt: new Date().toISOString()
  },
  {
    id: '2',
    name: 'Maria Santos',
    email: 'maria@example.com',
    createdAt: new Date().toISOString()
  }
]

function executeGraphQL(query: string, variables: Record<string, unknown>) {
  // GET_TASKS
  if (query.includes('query GetTasks')) {
    return {
      data: {
        tasks
      }
    }
  }

  // GET_USER
  if (query.includes('query GetUser')) {
    const user = users.find((u) => u.id === variables.id)
    return {
      data: {
        user: user || null
      }
    }
  }

  // CREATE_TASK
  if (query.includes('mutation CreateTask')) {
    const title = typeof variables.title === 'string' ? variables.title : ''
    const newTask = {
      id: String(tasks.length + 1),
      title,
      completed: false,
      createdAt: new Date().toISOString()
    }
    tasks.push(newTask)
    return {
      data: {
        createTask: newTask
      }
    }
  }

  // TOGGLE_TASK
  if (query.includes('mutation ToggleTask')) {
    const task = tasks.find((t) => t.id === variables.id)
    if (task) {
      task.completed = !task.completed
      return {
        data: {
          toggleTask: {
            id: task.id,
            completed: task.completed
          }
        }
      }
    }
    return {
      errors: [{ message: 'Task not found' }]
    }
  }

  // DELETE_TASK
  if (query.includes('mutation DeleteTask')) {
    const index = tasks.findIndex((t) => t.id === variables.id)
    if (index !== -1) {
      tasks.splice(index, 1)
      return {
        data: {
          deleteTask: {
            id: variables.id
          }
        }
      }
    }
    return {
      errors: [{ message: 'Task not found' }]
    }
  }

  return {
    errors: [{ message: 'Unknown query or mutation' }]
  }
}

export async function POST(request: NextRequest) {
  try {
    const body = await request.json()
    const { query, variables } = body

    const result = executeGraphQL(query, variables || {})

    return NextResponse.json(result)
  } catch {
    return NextResponse.json(
      {
        errors: [{ message: 'Invalid request' }]
      },
      { status: 400 }
    )
  }
}
