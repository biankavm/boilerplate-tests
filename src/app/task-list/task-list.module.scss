.container {
  max-width: 600px;
  margin: 0 auto;
  padding: 24px;
}

.title {
  font-size: 32px;
  font-weight: 600;
  margin-bottom: 24px;
  color: #333;
}

.loading,
.error,
.empty {
  padding: 32px;
  text-align: center;
  border-radius: 8px;
}

.loading {
  background-color: #e3f2fd;
  color: #1976d2;
}

.error {
  background-color: #ffebee;
  color: #d32f2f;
}

.empty {
  background-color: #f5f5f5;
  color: #666;
}

.emptySubtitle {
  font-size: 14px;
  margin-top: 8px;
  color: #999;
}

.list {
  display: flex;
  flex-direction: column;
}
