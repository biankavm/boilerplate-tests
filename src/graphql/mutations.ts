import { gql } from '@apollo/client'

export const CREATE_TASK = gql`
  mutation CreateTask($title: String!) {
    createTask(title: $title) {
      id
      title
      completed
      createdAt
    }
  }
`

export const TOGGLE_TASK = gql`
  mutation ToggleTask($id: ID!) {
    toggleTask(id: $id) {
      id
      completed
    }
  }
`

export const DELETE_TASK = gql`
  mutation DeleteTask($id: ID!) {
    deleteTask(id: $id) {
      id
    }
  }
`
