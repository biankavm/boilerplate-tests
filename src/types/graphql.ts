export interface Task {
  id: string
  title: string
  completed: boolean
  createdAt: string
}

export interface User {
  id: string
  name: string
  email: string
  createdAt: string
}

export interface GetTasksData {
  tasks: Task[]
}

export interface GetUserData {
  user: User | null
}

export interface GetUserVariables {
  id: string
}
