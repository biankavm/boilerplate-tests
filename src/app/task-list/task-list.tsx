'use client'

import { useQuery } from '@apollo/client/react'
import { GET_TASKS } from '@/graphql'
import { GetTasksData } from '@/types/graphql'
import { TaskItem } from '@/components/tasks'
import { CreateTask } from '@/components/tasks'
import styles from './task-list.module.scss'

export function TaskList() {
  const { data, loading, error } = useQuery<GetTasksData>(GET_TASKS)

  if (loading) {
    return (
      <div className={styles.container}>
        <h1 className={styles.title}>Minhas Tarefas</h1>
        <CreateTask />
        <div
          className={styles.loading}
          role="status"
          aria-label="Carregando tarefas"
        >
          <p>Carregando tarefas...</p>
        </div>
      </div>
    )
  }

  if (error) {
    return (
      <div className={styles.container}>
        <h1 className={styles.title}>Minhas Tarefas</h1>
        <CreateTask />
        <div className={styles.error} role="alert">
          <p>Erro ao carregar tarefas: {error.message}</p>
        </div>
      </div>
    )
  }

  const tasks = data?.tasks || []

  return (
    <div className={styles.container}>
      <h1 className={styles.title}>Minhas Tarefas</h1>
      <CreateTask />
      {tasks.length === 0 ? (
        <div className={styles.empty} role="status">
          <p>Nenhuma tarefa cadastrada ainda.</p>
          <p className={styles.emptySubtitle}>
            Comece criando sua primeira tarefa acima!
          </p>
        </div>
      ) : (
        <div className={styles.list} role="list">
          {tasks.map(
            (task: { id: string; title: string; completed: boolean }) => (
              <TaskItem
                key={task.id}
                id={task.id}
                title={task.title}
                completed={task.completed}
              />
            )
          )}
        </div>
      )}
    </div>
  )
}
