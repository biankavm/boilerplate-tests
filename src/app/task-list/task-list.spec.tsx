import { TaskList } from './task-list'
import { render, screen, waitFor } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing/react'
import { GET_TASKS } from '@/graphql'

const mocks = [
  {
    request: {
      query: GET_TASKS
    },
    result: {
      data: {
        tasks: [
          {
            id: '1',
            title: 'some title',
            completed: false,
            createdAt: '30/10/25'
          }
        ]
      }
    }
  }
]

const mocksNoTasks = [
  {
    request: {
      query: GET_TASKS
    },
    error: new Error('some error')
  }
]

jest.mock('@/components/tasks', () => ({
  TaskItem: jest.fn(({ id, title }) => (
    <div>
      {id} {title}
    </div>
  )),
  CreateTask: jest.fn(() => <div> create task test</div>)
}))

describe('TaskList', () => {
  const renderComp = () =>
    render(
      <MockedProvider mocks={mocks}>
        <TaskList />
      </MockedProvider>
    )

  it('should render correctly', async () => {
    const { container } = renderComp()

    expect(screen.getByText(/carregando tarefas.../i)).toBeInTheDocument()

    await waitFor(() => {
      expect(screen.getByText(/some title/i)).toBeVisible()
    })

    expect(screen.queryByText(/carregando tarefas.../i)).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('should render messages no tasks exists', async () => {
    render(
      <MockedProvider mocks={mocksNoTasks}>
        <TaskList />
      </MockedProvider>
    )

    // expect(
    //   await screen.findByText(/Nenhuma tarefa cadastrada ainda/i)
    // ).toBeInTheDocument()
    // expect(
    //   await screen.findByText(/Comece criando sua primeira tarefa acima!/i)
    // ).toBeInTheDocument()
    expect(
      await screen.findByText(/Erro ao carregar tarefas/i)
    ).toBeInTheDocument()
  })
})
