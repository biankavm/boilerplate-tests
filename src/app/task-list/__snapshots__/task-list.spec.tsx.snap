// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TaskList should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="title"
    >
      Minhas Tarefas
    </h1>
    <div>
       create task test
    </div>
    <div
      class="list"
      role="list"
    >
      <div>
        1
         
        some title
      </div>
    </div>
  </div>
</div>
`;
